packages/teuchos/comm/test/Comm/scatterv.cpp

```diff
@@ -0,0 +1,190 @@
+// @HEADER
+// *****************************************************************************
+//                    Teuchos: Common Tools Package
+//
+// Copyright 2004 NTESS and the Teuchos contributors.
+// SPDX-License-Identifier: BSD-3-Clause
+// *****************************************************************************
+// @HEADER
+
+#include "Teuchos_ConfigDefs.hpp"
+#include "Teuchos_DefaultComm.hpp"
+#include "Teuchos_CommHelpers.hpp"
+#ifdef HAVE_TEUCHOS_MPI
+#  include "Teuchos_DefaultMpiComm.hpp"
+#endif // HAVE_TEUCHOS_MPI
+#include "Teuchos_UnitTestHarness.hpp"
+
+
+template<class PacketType>
+bool
+testScatterv (bool& success, std::ostream& out,
+              const int root, const Teuchos::Comm<int>& comm)
+{
+  using Teuchos::scatterv;
+  using Teuchos::TypeNameTraits;
+  using std::endl;
+  typedef PacketType packet_type;
+
+  // Teuchos constructs the output stream such that it only prints on
+  // Process 0 anyway.
+  Teuchos::OSTab tab0 (out);
+  out << "Testing Teuchos::scatterv<int, "
+      << TypeNameTraits<packet_type>::name ()
+      << "> with root = " << root << endl;
+  Teuchos::OSTab tab1 (out);
+
+  int lclSuccess = 1;
+  int gblSuccess = lclSuccess;
+
+#ifdef HAVE_TEUCHOS_MPI
+  using Teuchos::MpiComm;
+  int errCode = MPI_SUCCESS;
+
+  const MpiComm<int>* mpiComm = dynamic_cast<const MpiComm<int>* > (&comm);
+  TEUCHOS_TEST_FOR_EXCEPTION
+    (mpiComm == NULL, std::logic_error, "Building with MPI, but default "
+     "communicator is not a Teuchos::MpiComm!");
```

Comment:
```suggestion
  const MpiComm<int>& mpiComm = Teuchos::dyn_cast<const MpiComm<int>> (comm);
```

Might using [`dyn_cast<>()`](https://docs.trilinos.org/dev/packages/teuchos/doc/html/group__teuchos__language__support__grp.html#ga83b8843eaab7fa2229416838a4776932)?